Refactor library and add SemanticVersion class (#10)
toitware/toit-semver: README.md

| `1.2.3-beta.01` > `1.2.3-beta` | Pre-release information is a set/array, parsed by `.`.  A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal. |
| `1.2.3-beta.2.1` > `1.2.3-beta.1` | If all of the preceding identifiers are equal, integers must be compared the normal way. |
| `1.2.3-beta.2` > `1.2.3-beta.1` | If all of the preceding identifiers are equal, integers must be compared the normal way. |
| `1.2.3-beta` > `1.2.3-1` | Where strings and integers must be compared, strings have a higher precedence than integers. |
| `1.2.3-beta+abcd` = `1.2.3-beta` | Build-metadata is not used when comparing. |
| `1.2.3-beta+sha.0beef` = `1.2.3-beta+sha.80081` | Build-metadata is not used when comparing. |

## Library Usage
#### Creating the object directly
Imports the library, and creates a `SemanticVersion` object directly.
```toit
import semver show *

main:
  // Instantiation by direct creation.
  semver-a := SemanticVersion 1 2 3

  // Prints '1.2.3'.

>   print $semver-a

Reviewer: Here and for the others.
```suggestion
  print semver-a
```

Author: Damn. wups. Corrected.

  // Prints 1.0.0-alpha.1+sha.23132.
  print $semver-d
```
#### Immutability
Semver objects are immutable, but there is an easy way to create a new instance with one or more changed properties:
```toit
  // Direct instantiation including pre-release.
  semver-e := SemanticVersion 1 2 3 --pre-releases=["alpha",1]

  // Prints '1.2.3-alpha.1'.
  print semver-e

  // Create semver-e-new with minor now = 15
  semver-e-new := semver-e.with --minor=15

  // Prints '1.15.3-alpha.1'.
  print semver-e-new
```
#### Object instantiation by string parsing
The library parses strings into a `SemanticVersion` object, which has methods
and functions.  This function respects the options that [relax parsing
rules](#modifying-parsing-rules).

Comparison operators are shown in the example below.
```toit
  string-f := "1.0.0"
  string-g := "1.0.0-beta.1"

  // Parse the strings into SemanticVersion objects.
  semver-f := SemanticVersion.parse string-f
  semver-g := SemanticVersion.parse string-g

  // Compare two objects: prints "f is later than g."
  if semver-f > semver-g:
    print "f is later than g."
  else:

>     print "g is later than f."

Reviewer: ```suggestion
    unreachable
```

Author: Removed similar examples and used the other helpers

  // Compare two objects: prints "f and g are different."
  if semver-f == semver-g:

>     print "f and g are the same."

Reviewer: ```suggestion
    unreachable
```

Author: Also section changed.

  else:
    print "f and g are different."

  // Prints 'null'

>   print semver-h:

Reviewer: ```suggestion
  print semver-h
```

Author: Nice.. thanks.  Corrected

```

#### Simple comparison using strings only
For convenience and backwards compatibility, it is also possible to compare
strings directly without creating the objects.

Similar to all `compare-to` functions the `compare` function returns -1 if the
left-hand side is less than the right-hand side; 0 if they are equal, and 1
otherwise.
```toit
  // Create strings
  v1 := "1.0.0"
  v1-beta := "1.0.0-beta.1"

  // Compare the two strings. Prints "Compare is: 1".
  print "Compare is: $(compare v1 v1-beta)"

  // Compare two strings: prints "Compare is: -1"
  print "Compare is: $(compare v1-beta v1)"

  // Compare two strings: prints "Compare is: 0"
  print "Compare is: $(compare v1 v1)"

> 

Reviewer: Add an example with metadata.

Author: Added.

```

## Modifying parsing rules
This library allows several options that relax some Semver rules.  These can be
used to prevent parsing failures that normally throw an error (which may
otherwise potentially cause code to crash/stop):
| Example flag | Result of using the flag |
| - | - |
| `.parse "1" --accept-missing-minor` |  Will parse as `1.0.0`. |
| `.parse "1.4" --accept-missing-minor` |  Will not parse.  Will parse with `--accept-missing-patch`. |
| `.parse "1.2" --accept-missing-patch` | Will parse as `1.2.0`. |
| `.parse "1.2.03" --accept-leading-zeros` | This, and other variations like `1.02.3`, `001.002.003` will all parse as `1.2.3`.  Also accepts leading zeros on numeric `--pre-releases`, which normally would not parse. |
| `.parse "v1.2.3" --accept-v` |  Ignores the preceeding 'v', and parses as `1.2.3`. |

> | `.parse "1.a.3" [--if-error]` | This, and other variations like `1.2.a` or `a.2.3`,  would normally fail parsing and throw an error. Using this argument will execute the supplied block in case of error.  (Search 'nullable' in this [Toit Documentation](https://docs.toit.io/language/objects-constructors-inheritance-interfaces).) |

Reviewer: Why "search for nullable" ?

Author: Before we changed to use blocks, having nullable variables wasn't something I was familiar with... so I thought to link to it.  Removed.

# Semantic Versioning Parser
A package with an object and functions for semantic versioning, compliant with
'Semantic Versioning 2.0.0'.  This library makes an *immutable* object available,
called `SemanticVersion`.

## What is semver?
> **Semantic Versioning** (semver) is a versioning system for software that
> provides a clear and predictable way to communicate changes in a codebase. It
> uses a three-part version number format: `MAJOR.MINOR.PATCH`.
>
> This system helps developers understand the nature of changes in
> a project - whether upgrading to a new version will break existing
> functionality or if the changes are simply enhancements or bug fixes.
>
> Semver is useful because it makes managing dependencies in software projects
> more predictable and transparent, allowing teams to handle upgrades, ensure
> compatibility, and reduce the risk of accidentally introducing breaking
> changes.

See: [semver.org](https://semver.org/).

## Semver Format/Definition
`x.y.z-a+b` where:
| Token | Name | Type | Notes |
| - | - | - | - |

> | `x` | `major` | integer | Denotes a major version, and is incremented when there are backward-incompatible changes. Part of 'version-core'. If unused, can be `0`. One of version-core must be non-zero. |

Reviewer: Having at least on version-core number equal to non-zero is limiting.
I wouldn't be against allowing `0.0.0` (and maybe adding a flag to check for that `--disallow-0`. hopefully with a better flag-name).

Author: Implmented the flag `--accept-version-core-zero` alongside the others.  Will add some tests and test before resolving. (Can reverse the default/logic if desired!)

> | `y` | `minor` | integer | Denotes a minor version differece, most often incremented on the introduction of backward-compatible feature additions. Part of 'version-core'. If unused, can be `0`. One of version-core must be non-zero. |

Reviewer: ```suggestion
| `y` | `minor` | integer | Denotes a minor version difference, most often incremented on the introduction of backward-compatible feature additions. Part of 'version-core'. If unused, can be `0`. One of version-core must be non-zero. |
```

others - please see: [semver.org](https://semver.org/) for the complete
reference.

### Formatting Examples
Examples of what can be seen in practical use:
- `0.0.0`: First official release — stable interface guaranteed moving forward.
- `3.0.0-alpha.1`: This would refer to unstable alpha version of v3.0.0. ".1"
  usually means first alpha of what will become version 3.0.0.
- `2.1.0-beta+exp.sha.5114f85`: This would denotes a beta build, and gives build
  metadata, likely showing a specific experiment/commit.

## String Parsing
One way to create a `SemanticVersion` object is to have the library parse a
string.  Parsing operates as shown in the table below:
| Example | Compliant? | Explanation | Parsing Result |
| - | - | - | - |
| `1.2.3`| Compliant | As per definition | :green_circle: Parses as is. \\n If these values are in separate variables, it is cheaper to [create directly](#creating-the-object-directly). |
| `1.2` | Not strictly compliant | Missed `patch` | :yellow_circle: Fails parsing but can be parsed with a switch. |
| `1` | Not strictly compliant | Misses `minor` and `patch` | :yellow_circle: Fails parsing, but can be parsed with a switch. |

> | `v1.2.3` | Not strictly compliant | Has a leading `v`. Acceptable in documentation. |  :yelow_circle: Fails, but can be parsed using a switch that drops the leading `v`. |

Reviewer: ```suggestion
| `v1.2.3` | Not strictly compliant | Has a leading `v`. Acceptable in documentation. |  :yellow_circle: Fails, but can be parsed using a switch that drops the leading `v`. |
```

| `1.02.3` | Not strictly compliant | Has a leading `0` in `minor` | :yellow_circle: Fails parsing, but will accept and drop leading `0`'s with a switch. |
|`1.0.0-beta`| Compliant | Accepted, but `pre-release` definition could be confusing. Suggest using `1.0.0-beta.1` | :green_circle: Parses successfully. |
| `1.0.0.1` | Not compliant | Uses four parts.  Last delimiter should be `-` or `+` to denote `pre-release` or `build-metadata` | :red_circle: Parsing fails. |

See [`tests`](https://github.com/toitware/toit-semver/tree/main/tests) folder and [`parse-test.toit`](https://github.com/toitware/toit-semver/tree/main/tests/parse-test.toit) for other cases and expected outcomes.

> ## Logical Operators

Reviewer: ```suggestion
## Comparison operators
```

> The library implements code to support the normal logic operators/comparators,

Reviewer: ```suggestion
The library implements code to support the standard comparators,
```

such as `>`,  `<=`, etc. The [standard](https://semver.org/) dictates rules
about these.  Not all are obvious at first.  They operate in the following way:
| Example | Explanation |
| - | - |
| `1.2.3` > `1.2.1` | As expected. |
| `1.2.3` > `1.2.3-beta` | Versions with pre-release information have a lower precedence than the same without pre-release information. [link](https://semver.org/#spec-item-9).
| `1.2.3-beta.01` > `1.2.3-beta` | Pre-release information is a set/array, parsed by `.`.  A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal. |

> | `1.2.3-beta.2.1` > `1.2.3-beta.1` | If all of the preceding identifiers are equal, integers must be compared the normal way. |

Reviewer: Change the higher ones to have a "10" instead of "2"? That makes it clear that it's not just a dump character-by-character comparison.

Author: Put the example and text detailing int vs lexical comparison.

| `1.2.3-beta.2` > `1.2.3-beta.1` | If all of the preceding identifiers are equal, integers must be compared the normal way. |
| `1.2.3-beta` > `1.2.3-1` | Where strings and integers must be compared, strings have a higher precedence than integers. |
| `1.2.3-beta+abcd` = `1.2.3-beta` | Build-metadata is not to be used when comparing. |

> | `1.2.3-beta+sha.0beef` = `1.2.3-beta+sha.80081` | Build-metadata is not to be used when comparing. |

Reviewer: ```suggestion
| `1.2.3-beta+abcd` = `1.2.3-beta` | Build-metadata is not used when comparing. |
| `1.2.3-beta+sha.0beef` = `1.2.3-beta+sha.80081` | Build-metadata is not used when comparing. |
```

## Library Usage
#### Creating the object directly:
Imports the library, and creates a `SemanticVersion` object directly.
```toit
import semver show *

main:
  // Instantiation by direct creation
  semver-a := SemanticVersion 1 2 3

  // Prints 1.2.3
  print "$semver-a"

  // Instantiation by direct creation, without 'minor'
  semver-b := SemanticVersion 1 2

  // Prints 1.2.0
  print "$semver-b"

  // Instantiation by direct creation, without 'minor' or 'patch'
  semver-c := SemanticVersion 1

  // Prints 1.0.0

>   print "$semver-c"

Reviewer: ```suggestion
  // Instantiation by direct creation.
  semver-a := SemanticVersion 1 2 3

  // Prints '1.2.3'.
  print "$semver-a"

  // Instantiation by direct creation, without 'minor'.
  semver-b := SemanticVersion 1 2

  // Prints '1.2.0'.
  print "$semver-b"

  // Instantiation by direct creation, without 'minor' or 'patch'.
  semver-c := SemanticVersion 1

  // Prints '1.0.0'.
  print "$semver-c"
```

```
#### Directly creating including pre-release:
Pre-release and build-metadata can also be specified directly.  These are held
as a list and must be specified as a list, even if only one element:
```toit
  // (Contiues from previous example)

  // Direct instantiation.
  semver-d := SemanticVersion 1 0 0 ["alpha","1"] ["sha",23132]

  // Prints 1.0.0-alpha.1+sha.23132
  print "$semver-d"
```
#### Immutability:
Since the object is immutable, editing one of the fields after creation is not
possible.
```toit
  // (Contiues from previous examples)

  // Direct instantiation including pre-release.
  semver-e := SemanticVersion 1 2 3 ["alpha",1]

  // Prints 2
  print "$(semver-e.minor)"

  // Fails/throws
  //semver-e.minor = 5

> ```

Reviewer: I would refactor this:
If there isn't yet (haven't looked at the code yet):
Add a `with` function:
```
with --major/int?=null --minor/int?=null --patch/int?=null --pre-rerease/List?=null --build-metadata/List?=null:
  return SemanticVersion
      major or this.major
      minor or this.minor
      ...
```
You could then start by saying that the objects are immutable, but that there is an easy way to adjust individual segments.

Author: `with` function added, example in README extended

Author: Done

#### Object instantiation by string parsing:
The library parses strings into a `SemanticVersion` object, which has methods
and functions.  Comparison operators are shown in the example below.
```toit

>   // (Contiues from previous examples)

Reviewer: ```suggestion
  // (Continues from previous examples).
```

>   // strings

Reviewer: ```suggestion
```

  string-f := "1.0.0"
  string-g := "1.0.0-beta.1"

>   // parse strings into SemanticVersion objects

Reviewer: ```suggestion
  // Parse the strings into SemanticVersion objects.
```

  semver-f := SemanticVersion.parse string-f
  semver-g := SemanticVersion.parse string-g

  // compare two objects: prints "f is later than g."
  if semver-f > semver-g:
    print "f is later than g."
  else:
    print "g is later than f."

  // compare two objects: prints "f and g are different."
  if semver-f == semver-g:
    print "a and b are the same."
  else:
    print "a and b are different."

  // compare two objects: prints "f and f are the same."
  if semver-f == semver-f:
    print "f and f are the same."
  else:
    print "f and f are different."

> 

Reviewer: Not needed.
```suggestion
```

Consider adding an example for the yellow strings, where we accept them with specific flags.
In that case, it's also interesting to know whether a stringification/to-string returns the original string or a correct one.
So does `(SemanticVersion.parse "1.1").stringify` return "1.1" or "1.1.0"?

```

#### Simple comparison using strings only:
This is implemented, but not recommended as it is a computationally expensive -
string parsing happens every time an evaluation is made. (Code uses the library

> and creates/destroys objects in the background):

Reviewer: ```suggestion
For convenience and backwards compatibility, it is also possible to compare strings directly. In
the background the library creates the corresponding `SemanticVersion` instances and uses
them for comparison.

Similar to all `compare-to` functions the `compare` function returns -1 if the left-hand side is less
than the right-hand side; 0 if they are equal, and 1 otherwise.
```

```toit

>   // Contiues from previous examples

Reviewer: ```suggestion
  // Continues from previous examples
```

  // Create strings
  h := "1.0.0"

>   i := "1.0.0-beta.1"
> 
>   // compare two strings: prints "Compare is: 1"
>   print "Compare is: $(compare h i)"
> 
>   // compare two strings: prints "Compare is: -1"
>   print "Compare is: $(compare i h)"
> 
>   // compare two strings: prints "Compare is: 0"
>   print "Compare is: $(compare i i)"
> 

Reviewer: ```suggestion
  // Create strings
  v1 := "1.0.0"
  v1-beta := "1.0.0-beta.1"

  // Compare the two strings. Prints "Compare is: 1".
  print "Compare is: $(compare v1 v1-beta)"

  // compare two strings: prints "Compare is: -1"
  print "Compare is: $(compare v1-beta v1)"

  // compare two strings: prints "Compare is: 0"
  print "Compare is: $(compare v1 v1)"

```

```

## Switches relaxing some parsing rules
This library allows several switches that relax some of the rules in order to
prevent parsing failures that normally throw an error (eg, potentially causing
code to crash/stop):
| Example switch | Examples using the switch |
| - | - |
| `.parse "1" --accept-missing-minor` |  Will parse as `1.0.0`.  (Also assumes `--accept-missing-patch`) `1.2` will fail without the next switch. |
| `.parse "1.2" --accept-missing-patch` | Will parse as `1.2.0` |
| `.parse "1.2.03" --accept-leading-zeros` | This, and other variations like `1.02.3`, `001.002.003` will all parse as `1.2.3` |
| `.parse "v1.2.3" --accept-v` |  parses as `1.2.3` |

> | `.parse "1.a.3" --non-throwing` | This, and other variations like `1.2.a` or `a.2.3`,  would normally fail parsing and throw an error. Using this switch will generate `null` instead. (Search 'nullable' in this [Toit Documentation](https://docs.toit.io/language/objects-constructors-inheritance-interfaces)) |

Reviewer: This should be changed to `parse [--if-error]`. That's the standard for all our parsing functions.

Author: Done..

Author: Clarified in comments.

#### Example Combinations and Exceptions :
Switches can work in combination, as per the following examples:
| Example combination | Result |
| - | - |

> | `.parse "1" --accept-missing-patch` | Throws.  The switch only excuses missing `patch`, eg still expects `minor`. |

Reviewer: ```suggestion
| `.parse "1" --accept-missing-patch` | Throws.  The switch only excuses missing `patch`, but still expects `minor`. |
```

> | `.parse "1" --accept-missing-patch --non-throwing` | Even with both switches this will return `null`. The `--accept-missing-patch` switch only excepts a missing `patch` definition.  In this case it would still expect the `minor` integer. |

Reviewer: ```suggestion
| `.parse "1" --accept-missing-patch --if-error=(: null)` | Invokes the `if-error` block since a `minor` is missing, thus returning `null`. |
```

> | `.parse 1.54` | In this case 1.2 is a `float` and will return `1.2.0`, assuming `0` for the patch value.  **Note:** If the `float` is the result of some earlier math and is stored as `1.539999999997`, intervention is required if the intention was `1.54`. |

Reviewer: I would not accept floats. Parsing is done from strings. If necessary, users can just `parse "$version"`, but it's a really bad idea to go through floats.
for example: `1.10` will become `1.1`. We should not encourage version numbers from float.

Author: Good call. I'll remove it. :)

Author: Float capability removed.

```toit
import semver show *

main:
  // Instantiation by direct creation.
  semver-a := SemanticVersion 1 2 3

  // Prints '1.2.3'.
  print semver-a

  // Instantiation by direct creation, without 'minor'.
  semver-b := SemanticVersion 1 2

  // Prints '1.2.0'.
  print semver-b

  // Instantiation by direct creation, without 'minor' or 'patch'.
  semver-c := SemanticVersion 1

  // Prints '1.0.0'.
  print semver-c
```
#### Directly creating including pre-release
Pre-release and build-metadata can also be specified directly.  These are held
as a list and must be specified as a list, even if only one element:
```toit
  // Direct instantiation.
  semver-d := SemanticVersion 1 0 0 --pre-releases=["alpha","1"] --build-metadata=["sha",23132]

  // Prints 1.0.0-alpha.1+sha.23132.
  print semver-d
```
#### Immutability
Semver objects are immutable, but there is an easy way to create a new instance with one or more changed properties:
```toit
  // Direct instantiation including pre-release.
  semver-e := SemanticVersion 1 2 3 --pre-releases=["alpha",1]

  // Prints '1.2.3-alpha.1'.
  print semver-e

  // Create semver-e-new with minor now = 15

>   semver-e-new := semver-e.with --minor=15

Reviewer: for a follow-up PR, it could be nice to have `.next-patch`, `.next-minor`, and `.next-major`.
It feels like these are the only real uses-cases in the wild.
Again: not this PR.

Author: Ok. Leaving this open for now till its time.

Author: Copied this to the final comment, to track additional changes in a new PR.  Closing here.

  // Prints '1.15.3-alpha.1'.
  print semver-e-new
```
#### Object instantiation by string parsing
The library parses strings into a `SemanticVersion` object, which has methods
and functions.  This function respects the options that [relax parsing
rules](#modifying-parsing-rules).

Comparison operators are shown in the example below.
```toit
  string-f := "1.0.0"
  string-g := "v3.10.1-beta.1"

  // Parse the strings into SemanticVersion objects.
  semver-f := SemanticVersion.parse string-f
  semver-g := SemanticVersion.parse string-g --accept-v

  // prints "1.0.0".
  print semver-f

  // prints "3.10.1-beta.1".
  // (Note that the v was dropped during parsing.)
  print semver-g
```

#### Simple comparison
Similar to all `compare-to` functions the `compare` function returns -1 if the
left-hand side is less than the right-hand side; 0 if they are equal, and 1
otherwise.  `precedes` and `equals` produce boolean results.
```toit
  // Parse the strings into SemanticVersion objects.
  semver-h := SemanticVersion 1 20 3
  semver-i := SemanticVersion 2 5 10

  // Compare two objects with output: prints "1.20.3 compare-to 2.5.10 is -1."

>   print "$semver-h compare-to $semver-i is $((semver-h.compare-to semver-i))."

Reviewer: ```suggestion
  print "$semver-h compare-to $semver-i is $(semver-h.compare-to semver-i)."
```

Similar to all `compare-to` functions the `compare` function returns -1 if the
left-hand precedes the right-hand side; 0 if they are equal, and 1
otherwise.
```toit
  // Create strings
  str-core := "1.0.0"
  str-pre-release := "1.0.0-beta.1"
  str-build-metadata := "1.0.0-beta.1+build.82f4c8f"

  // Compare the two strings. Prints "Compare is: 1".
  print "Compare is: $(compare str-core str-pre-release)"

  // Compare two strings: prints "Compare is: -1"
  print "Compare is: $(compare str-pre-release str-core)"

  // Compare two strings: prints "Compare is: 0"
  print "Compare is: $(compare str-pre-release str-pre-release)"

  str-build-metadata-2 := "1.0.0-beta.1+build.f72ae1a"

  // Compare two strings: prints "Compare is: 0"
  print "Compare is: $(compare str-build-metadata str-build-metadata-2)"
```

## Modifying parsing rules
This library allows several options that relax some Semver rules.  These can be
used to prevent parsing failures that normally throw an error (which may
otherwise potentially cause code to crash/stop):
| Example flag | Result of using the flag |
| - | - |
| `.parse "1" --accept-missing-minor` |  Will parse as `1.0.0`. |

> | `.parse "1.4" --accept-missing-minor` |  Will not parse.  Will parse with `--accept-missing-patch`. |

Reviewer: I think `accept-missing-minor` should imply `accept-missing-patch`.

Author: Your call! Parser constructor adjusted to enforce this. [305217d](https://github.com/toitware/toit-semver/pull/10/commits/305217d498b5041bd1094fc5a0cc0d3dc9d95a58)

> provides a clear and predictable way to communicate changes in a codebase. It
> uses a three-part version number format: `MAJOR.MINOR.PATCH`.
>
> This system helps developers understand the nature of changes in
> a project - whether upgrading to a new version will break existing
> functionality or if the changes are simply enhancements or bug fixes.
>
> Semver is useful because it makes managing dependencies in software projects
> more predictable and transparent, allowing teams to handle upgrades, ensure
> compatibility, and reduce the risk of accidentally introducing breaking
> changes.

See: [semver.org](https://semver.org/).

## Semver Format/Definition
`x.y.z-a+b` where:
| Token | Name | Type | Notes |
| - | - | - | - |
| `x` | `major` | integer | Denotes a major version, and is incremented when there are backward-incompatible changes. Part of 'version-core'. If unused, can be `0`. One of version-core must be non-zero. |
| `y` | `minor` | integer | Denotes a minor version difference, most often incremented on the introduction of backward-compatible feature additions. Part of 'version-core'. If unused, can be `0`. One of version-core must be non-zero. |

> | `z` | `patch` | integer | **MANDATORY:**Incremented for backward-compatible fixes or improvements. Part of 'version-core'. If unused, can be `0`. One of version-core must be non-zero.|

Reviewer: ```suggestion
| `z` | `patch` | integer | **MANDATORY:** Incremented for backward-compatible fixes or improvements. Part of 'version-core'. If unused, can be `0`. One of version-core must be non-zero.|
```

| `a` | `pre-release` | Set of zero or more integers/strings, delimited by `.` | **OPTIONAL:** Indicates the version is a pre-release.  The same version without the pre-releases will be ahead (newer/better/greater than) the pre-release version. |
| `b` | `build-metadata` | Set of zero or more integers/strings, delimited by `.` | **OPTIONAL:** This portion is informational only, and usually references a specific build.  The semver standard dictates that it MUST be ignored when determining version precedence. [link](https://semver.org/#spec-item-10). |

There are more detailed rules around what characters are valid after certain
others - please see: [semver.org](https://semver.org/) for the complete
reference.

### Formatting Examples
Examples of what can be seen in practical use:

> - `0.0.0`: First official release — stable interface guaranteed moving forward.

Reviewer: according to the doc above, 0.0.0 wouldn't even be valid.
```suggestion
- `1.0.0`: First official release — stable interface guaranteed moving forward.
```

Author: Done - thanks

- `3.0.0-alpha.1`: This would refer to unstable alpha version of v3.0.0. ".1"
  usually means first alpha of what will become version 3.0.0.
- `2.1.0-beta+exp.sha.5114f85`: This would denotes a beta build, and gives build
  metadata, likely showing a specific experiment/commit.

## String Parsing
One way to create a `SemanticVersion` object is to have the library parse a
string.  Parsing operates as shown in the table below:
| Example | Compliant? | Explanation | Parsing Result |
| - | - | - | - |

> | `1.2.3` | Compliant | As per definition. | :green_circle: Parses as is. \\n If these values are in separate variables, it is cheaper to [create directly](#creating-the-object-directly). |

Reviewer: ```suggestion
| `1.2.3` | Compliant | As per definition. | :green_circle: Parses as is. If the individual tokens (major, minor, patch) are already separated, it is cheaper to [create an object directly](#creating-the-object-directly). |
```

Author: Done

| `1.2` | Not compliant | Missing `patch`. | :yellow_circle: Fails parsing but can be parsed with a [switch](#switches-relaxing-some-parsing-rules). |
| `1` | Not compliant | Missing `minor` and `patch`. | :yellow_circle: Fails parsing, but can be parsed with a [switch](#switches-relaxing-some-parsing-rules). |
| `v1.2.3` | Not compliant | Has a leading `v`. Acceptable in documentation. |  :yellow_circle: Fails, but can be parsed using a [switch](#switches-relaxing-some-parsing-rules) that ignores the leading `v`. |
| `1.02.3` | Not compliant. | Has a leading `0` in `minor`. | :yellow_circle: Fails parsing, but will accept and drop leading `0`'s with a [switch](#switches-relaxing-some-parsing-rules). |

> | `1.0.0-beta` | Compliant. | Accepted, but `pre-release` definition could be confusing. Suggest using `1.0.0-beta.1` | :green_circle: Parses successfully. |

Reviewer: ```suggestion
| `1.0.0-beta` | Compliant. | Accepted.| :green_circle: Parses successfully. |
| `1.0.0-beta.1` | Compliant. | Accepted.| :green_circle: Parses successfully. |
```

Author: Done

| `1.0.0.1` | Not compliant. | Use of four parts is not compliant.  If a pre-release/build-metadata was intended, the last delimiter should have been `-` or `+` to denote `pre-release` or `build-metadata` | :red_circle: Parsing fails. |

See [`tests`](https://github.com/toitware/toit-semver/tree/main/tests) folder and [`parse-test.toit`](https://github.com/toitware/toit-semver/tree/main/tests/parse-test.toit) for other cases and expected outcomes.

## Comparison operators
The library implements code to support the standard comparators,

> such as `>`,  `<=`, etc. The [standard](https://semver.org/) dictates rules

Reviewer: I wonder if this is a good idea.
`1.2.3+build1` shouldn't be equal to `1.2.3+build2`. But if the operators are implemented, then `1.2.3+build1 <= 1.2.3+build2` but not `<`.
So we have a situation, where a <= b, but neither a < b nor a == b.

Author: Added an argument to `compare-to` which evaluates the build-metadata too: `.compare-to other --compare-build-metadata=false -> int` and is off by default.  I think this is dangerous to leave on by default, because allegedly, build metadata most often contains those commit identifiers etc, which may or may not be sequential.  Maybe `equals` compares `build-metadata` too, but the other comparisons don't?

Reviewer: As discussed in chat: an `a == b` should imply that it prints the same way. So `==` should include metadata.
However, we want to add `equals` and `precedes`, and remove the other operators.

This section now needs a rewrite.

Author: Rewritten.. operators removed, and helper functions added as above.

about these.  Not all are obvious at first.  They operate in the following way:
| Example | Explanation |
| - | - |
| `1.2.3` > `1.2.1` | As expected. |
| `1.2.3` > `1.2.3-beta` | Versions with pre-release information have a lower precedence than the same without pre-release information. [link](https://semver.org/#spec-item-9).
| `1.2.3-beta.01` > `1.2.3-beta` | Pre-release information is a set/array, parsed by `.`.  A larger set of pre-release fields has a higher precedence than a smaller set, if all of the preceding identifiers are equal. |
| `1.2.3-beta.2.1` > `1.2.3-beta.1` | If all of the preceding identifiers are equal, integers must be compared the normal way. |
| `1.2.3-beta.2` > `1.2.3-beta.1` | If all of the preceding identifiers are equal, integers must be compared the normal way. |
| `1.2.3-beta` > `1.2.3-1` | Where strings and integers must be compared, strings have a higher precedence than integers. |
| `1.2.3-beta+abcd` = `1.2.3-beta` | Build-metadata is not used when comparing. |
| `1.2.3-beta+sha.0beef` = `1.2.3-beta+sha.80081` | Build-metadata is not used when comparing. |

## Library Usage

> #### Creating the object directly:

Reviewer: ```suggestion
#### Creating the object directly
```

Author: Done

Imports the library, and creates a `SemanticVersion` object directly.
```toit
import semver show *

main:
  // Instantiation by direct creation.
  semver-a := SemanticVersion 1 2 3

  // Prints '1.2.3'.

>   print "$semver-a"

Reviewer: ```suggestion
  print semver-a
```

Author: Done and others checked.

  // Instantiation by direct creation, without 'minor'.
  semver-b := SemanticVersion 1 2

  // Prints '1.2.0'.

>   print "$semver-b"

Reviewer: ```suggestion
  print semver-b
```

Author: Done

  // Instantiation by direct creation, without 'minor' or 'patch'.
  semver-c := SemanticVersion 1

  // Prints '1.0.0'.

>   print "$semver-c"

Reviewer: ```suggestion
  print semver-c
```

Author: Done

```

> #### Directly creating including pre-release:

Reviewer: ```suggestion
#### Directly creating including pre-release
```

Author: Done

Pre-release and build-metadata can also be specified directly.  These are held
as a list and must be specified as a list, even if only one element:
```toit
  // (Continues from previous example)

  // Direct instantiation.
  semver-d := SemanticVersion 1 0 0 --pre-releases=["alpha","1"] --build-metadata=["sha",23132]

>   // Prints 1.0.0-alpha.1+sha.23132

Reviewer: ```suggestion
  // Prints 1.0.0-alpha.1+sha.23132.
```

Author: Done

  print "$semver-d"
```

> #### Immutability:

Reviewer: ```suggestion
#### Immutability
```

Author: Done

Since the object is immutable, editing one of the fields after creation is not
possible.
```toit
  // (Continues from previous examples.)

  // Direct instantiation including pre-release.
  semver-e := SemanticVersion 1 2 3 --pre-releases=["alpha",1]

  // Prints '1.2.3-alpha.1'
  print "$(semver-e)"

  // Prints 2
  print "$(semver-e.minor)"

  // Fails/throws
  //semver-e.minor = 15
```
However, there is an easy way to quickly create a new instance with one or
more changed properties:
```toit
  // (Continues from previous examples.)

  // Create semver-e-new with minor now = 15
  semver-e-new := semver-e.with --minor=15

  // Prints '1.15.3-alpha.1'

>   print "$(semver-e-new)"

Reviewer: ```suggestion
Semver objects are immutabel, but there is an easy way to create a new instance with one or more changed properties:

```toit
  // Direct instantiation including pre-release.
  semver-e := SemanticVersion 1 2 3 --pre-releases=["alpha",1]

  // Prints '1.2.3-alpha.1'.
  print semver-e

  // Create semver-e-new with minor now = 15
  semver-e-new := semver-e.with --minor=15

  // Prints '1.15.3-alpha.1'.
  print semver-e-new
```

Author: Done

```

> #### Object instantiation by string parsing:

Reviewer: ```suggestion
#### Object instantiation by string parsing
```

Author: All colons taken out of those L4 headings.

The library parses strings into a `SemanticVersion` object, which has methods
and functions.  Comparison operators are shown in the example below.
```toit
  // (Continues from previous examples.).

  string-f := "1.0.0"
  string-g := "1.0.0-beta.1"

  // Parse the strings into SemanticVersion objects.
  semver-f := SemanticVersion.parse string-f
  semver-g := SemanticVersion.parse string-g

>   // compare two objects: prints "f is later than g."

Reviewer: ```suggestion
  // Compare two objects: prints "f is later than g."
```

Author: Adjusted

>   if semver-f > semver-g:

Reviewer: As written above: not sure if we shouldn't just remove the operators and only keep `compare-to`.

Author: Done - resolving this and handling elsewhere.

    print "f is later than g."
  else:
    print "g is later than f."

>   // compare two objects: prints "f and g are different."

Reviewer: ```suggestion
  // Compare two objects: prints "f and g are different."
```

Author: Done

  if semver-f == semver-g:
    print "f and g are the same."
  else:
    print "f and g are different."

>   // compare two objects: prints "f and f are the same."

Reviewer: ```suggestion
  // Compare two objects: prints "f and f are the same."
```

Author: Done

  if semver-f == semver-f:
    print "f and f are the same."
  else:
    print "f and f are different."
```
If a string does not parse, it will throw an error unless `--if-error` block is
supplied, as per this example:
```toit
  // (Continues from previous examples.).

  // Has 4 parts to the version-core, will not parse.
  string-h := "1.0.0.5-beta.1"

>   // Parses with the --if-error switch.  Will not throw but execute/return the

Reviewer: Usually, we don't use the term "switch" in Toit.

```suggestion
  // Parses with the --if-error block.  Will not throw but execute/return the
```

Author: Removed.

Author: Had a crack at changing the entire readme.

  // block when parsing fails.
  semver-h := SemanticVersion.parse string-h --if-error=(: null)

  // Prints 'is null'
  if not semver-h:

>     print "is null"

Reviewer: ```suggestion
  // Prints 'null'
  print semver-h
```

Author: Done

```

> #### Simple comparison using strings only:

Reviewer: ```suggestion
#### Simple comparison using strings only
```

Author: Done

For convenience and backwards compatibility, it is also possible to compare
strings directly without creating the objects. In the background the library
creates the corresponding `SemanticVersion` object anyway, and uses the built-in
functions/methods for comparison.

Similar to all `compare-to` functions the `compare` function returns -1 if the
left-hand side is less than the right-hand side; 0 if they are equal, and 1
otherwise.
```toit
  // (Continues from previous examples.)

> 

Reviewer: There might be other places where we can remove this.
```suggestion
```

Author: Searched and removed all.

This library allows several switches that relax some of the rules in order to
prevent parsing failures that normally throw an error (which may otherwise
potentially cause code to crash/stop):
| Example switch | Examples using the switch |
| - | - |
| `.parse "1" --accept-missing-minor` |  Will parse as `1.0.0`. |
| `.parse "1.4" --accept-missing-minor` |  Will not parse.  Will parse with `--accept-missing-patch`. |
| `.parse "1.2" --accept-missing-patch` | Will parse as `1.2.0`. |
| `.parse "1.2.03" --accept-leading-zeros` | This, and other variations like `1.02.3`, `001.002.003` will all parse as `1.2.3`.  Also accepts leading zeros on numeric `--pre-releases`, which normally would not parse. |
| `.parse "v1.2.3" --accept-v` |  Ignores the preceeding 'v', and parses as `1.2.3`. |
| `.parse "1.a.3" --if-error=(: null)` | This, and other variations like `1.2.a` or `a.2.3`,  would normally fail parsing and throw an error. Using this switch will execute the supplied block instead, in this case producing `null` as the result.  (Search 'nullable' in this [Toit Documentation](https://docs.toit.io/language/objects-constructors-inheritance-interfaces).) |

> #### Example Combinations and Exceptions :

Reviewer: ```suggestion
#### Example Combinations and Exceptions
```

Author: Done
